src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,191 @@
+                containerIdAndRoot.Item1,
+                projectId,
+                containerIdAndRoot.Item2
+                );
+
+            tracer.Info($"Start downloading FCS artifact- {artifact.Name}");
+            IEnumerable<Func<string, bool>> minimatcherFuncs = MinimatchHelper.GetMinimatchFuncs(minimatchPatterns, tracer);
+            if (minimatcherFuncs !=null && minimatcherFuncs.Count() !=0)
+            {
+                items = this.GetFilteredItems(items, minimatcherFuncs, containerIdAndRoot.Item2);
+            }
+
+            if(!isSingleArtifactDownload && items.Any())
+            {
+                Directory.CreateDirectory(rootPath);
+            }
+
+            var folderItems = items.Where(i => i.ItemType == ContainerItemType.Folder);
+            Parallel.ForEach(folderItems, (folder) =>
+            {
+                var targetPath = ResolveTargetPath(rootPath, folder, containerIdAndRoot.Item2);
+                Directory.CreateDirectory(targetPath);
+            });
+
+            var fileItems = items.Where(i => i.ItemType == ContainerItemType.File);
+
+            var downloadBlock = NonSwallowingActionBlock.Create<FileContainerItem>(
+                async item =>
+                {
+                    var targetPath = ResolveTargetPath(rootPath, item, containerIdAndRoot.Item2);
+                    var directory = Path.GetDirectoryName(targetPath);
+                    Directory.CreateDirectory(directory);
+                    await AsyncHttpRetryHelper.InvokeVoidAsync(
+                       async () =>
+                       {
+                           using (var sourceStream = await this.DownloadFileFromContainerAsync(containerIdAndRoot, projectId, containerClient, item, cancellationToken))
+                           {
+                               tracer.Info($"Downloading: {targetPath}");
+                               using (var targetStream = new FileStream(targetPath, FileMode.Create))
+                               {
+                                   sourceStream.CopyTo(targetStream);
+                               }
+                           }
+                       },
+                        maxRetries: 5,
+                        cancellationToken: cancellationToken,
+                        tracer: tracer,
+                        continueOnCapturedContext: false,
+                        canRetryDelegate: exception => exception is IOException,
+                        context: null
+                        );
+                },
+                new ExecutionDataflowBlockOptions()
+                {
+                    BoundedCapacity = 5000,
+                    MaxDegreeOfParallelism = 8,
+                    CancellationToken = cancellationToken,
+                });
+
+                await downloadBlock.SendAllAndCompleteSingleBlockNetworkAsync(fileItems, cancellationToken);
+        }
+
+        private async Task<Stream> DownloadFileFromContainerAsync(
```

Comment:
>DownloadFileFromContainerAsync [](start = 35, length = 30)

This is not downloading, rather it's fetching the stream

Reply:
Yeah, but as we are calling the "DownloadFileAsync" FCS function, named it as DownloadFileFromContainerAsync. We can rename it.

Reply:
Keep it the same then